packages/expo-file-system/android/src/main/java/expo/modules/filesystem/FileSystemModule.kt

```diff
@@ -0,0 +1,1377 @@
+      val capacityDouble = Math.min(capacity.toDouble(), 2.0.pow(53.0) - 1)
+      promise.resolve(capacityDouble)
+    } catch (e: Exception) {
+      e.message?.let { Log.e(TAG, it) }
+      promise.reject("ERR_FILESYSTEM_CANNOT_DETERMINE_DISK_CAPACITY", "Unable to access total disk capacity", e)
+    }
+  }
+
+  @ExpoMethod
+  fun getFreeDiskStorageAsync(promise: Promise) {
+    try {
+      val external = StatFs(Environment.getDataDirectory().absolutePath)
+      val availableBlocks = external.availableBlocksLong
+      val blockSize = external.blockSizeLong
+      val storage = BigInteger.valueOf(availableBlocks).multiply(BigInteger.valueOf(blockSize))
+      // cast down to avoid overflow
+      val storageDouble = Math.min(storage.toDouble(), 2.0.pow(53.0) - 1)
+      promise.resolve(storageDouble)
+    } catch (e: Exception) {
+      e.message?.let { Log.e(TAG, it) }
+      promise.reject("ERR_FILESYSTEM_CANNOT_DETERMINE_DISK_CAPACITY", "Unable to determine free disk storage capacity", e)
+    }
+  }
+
+  @ExpoMethod
+  fun getContentUriAsync(uri: String, promise: Promise) {
+    try {
+      val fileUri = Uri.parse(uri)
+      ensurePermission(fileUri, Permission.WRITE)
+      ensurePermission(fileUri, Permission.READ)
+      checkIfFileDirExists(fileUri)
+      if (fileUri.scheme == "file") {
+        val file = uriToFile(fileUri)
+        promise.resolve(contentUriFromFile(file).toString())
+      } else {
+        promise.reject("ERR_FILESYSTEM_CANNOT_READ_DIRECTORY", "No readable files with the uri: $uri. Please use other uri.")
+      }
+    } catch (e: Exception) {
+      e.message?.let { Log.e(TAG, it) }
+      promise.reject(e)
+    }
+  }
+
+  private fun contentUriFromFile(file: File): Uri {
+    return try {
+      val activityProvider: ActivityProvider by moduleRegistry()
+      val application = activityProvider.currentActivity.application
+      FileProvider.getUriForFile(application, application.packageName + ".FileSystemFileProvider", file)
+    } catch (e: Exception) {
+      throw e
+    }
+  }
+
+  @ExpoMethod
+  fun readSAFDirectoryAsync(uriStr: String?, options: Map<String?, Any?>?, promise: Promise) {
+    try {
+      val uri = Uri.parse(uriStr)
+      ensurePermission(uri, Permission.READ)
+      if (isSAFUri(uri)) {
+        val file = DocumentFile.fromTreeUri(context, uri)
+        if (file == null || !file.exists() || !file.isDirectory) {
+          promise.reject(
+            "ERR_FILESYSTEM_CANNOT_READ_DIRECTORY",
+            "Uri '$uri' doesn't exist or isn't a directory."
+          )
+          return
+        }
+        val children = file.listFiles()
+        val result: MutableList<String> = ArrayList()
+        for (child in children) {
+          result.add(child.uri.toString())
+        }
+        promise.resolve(result)
+      } else {
+        throw IOException("The URI '$uri' is not a Storage Access Framework URI. Try using FileSystem.readDirectoryAsync instead.")
```

Comment:
Not really subject of this PR, but this is a good opportunity to unify error messages:
- To put method names in quotes or not (`FileSystem.readDirectoryAsync` vs `'FileSystem.readDirectoryAsync'`)
- To add brackets or not (`FileSystem.readDirectoryAsync` vs `FileSystem.readDirectoryAsync()`)